processing/src/main/java/org/apache/druid/segment/IndexMergerV9.java

```diff
@@ -335,7 +263,7 @@ private File makeIndexFiles(
           // shouldStore AND hasOnlyNulls
           ColumnDescriptor columnDesc = ColumnDescriptor
               .builder()
-              .setValueType(dimCapabilities.get(i).getType())
+              .setValueType(dimFormats.get(i).getLogicalType().getType())
               .addSerde(new NullColumnPartSerde(indexMergeResult.rowCount, indexSpec.getBitmapSerdeFactory()))
               .build();
```

Comment:
this is from that explicit null value columns change we did a while back that saves any dimension schema in the segment that didn't have any values

I agree this can probably be better/smarter